docs/src/pages/docs/guides/migrating-to-react-query-3.md

```diff
@@ -0,0 +1,241 @@
+---
+id: migrating-to-react-query-3
+title: Migrating to React Query 3
+---
+
+## V3 migration
+
+This article explains how to migrate your application to React Query 3.
+
+### QueryClient
+
+The `QueryCache` has been split into a `QueryClient` and a `QueryCache`.
+The `QueryCache` contains all cached queries and the `QueryClient` can be used to interact with a cache.
+
+This has some benefits:
+
+- Allows for different type of caches.
+- Multiple clients with different configurations can use the same cache.
+- Clients can be used to track queries, which can be used for shared caches on SSR.
+- The client API is more focused towards general usage.
+- Easier to test the individual components.
+
+Use the `QueryClientProvider` component to connect a `QueryClient` to your application:
+
+```js
+import { QueryClient, QueryClientProvider, QueryCache } from 'react-query'
+
+const cache = new QueryCache()
+const client = new QueryClient({ cache })
+
+function App() {
+  return <QueryClientProvider client={client}>...</QueryClientProvider>
+}
+```
+
+### useQueryCache()
+
+The `useQueryCache()` hook has been replaced by the `useQueryClient()` hook:
+
+```js
+import { useCallback } from 'react'
+import { useQueryClient } from 'react-query'
+
+function Todo() {
+  const client = useQueryClient()
+
+  const onClickButton = useCallback(() => {
+    client.refetchQueries('posts')
+  }, [client])
+
+  return <button onClick={onClickButton}>Refetch</button>
+}
+```
+
+### ReactQueryConfigProvider
+
+The `ReactQueryConfigProvider` component has been removed. Default options for queries and mutations can now be specified in `QueryClient`:
+
+```js
+const client = new QueryClient({
+  cache,
+  defaultOptions: {
+    queries: {
+      staleTime: Infinity,
+    },
+  },
+})
+```
+
+### usePaginatedQuery()
+
+The `usePaginatedQuery()` hook has been replaced by the `keepPreviousData` option on `useQuery`:
+
+```js
+import { useQuery } from 'react-query'
+
+function Page({ page }) {
+  const { data } = useQuery(['page', page], fetchPage, {
+    keepPreviousData: true,
+  })
+}
+```
+
+### Query object syntax
+
+The object syntax has been collapsed:
+
+```js
+// Old:
+useQuery({
+  queryKey: 'posts',
+  queryFn: fetchPosts,
+  config: { staleTime: Infinity },
+})
+
+// New:
+useQuery({
+  queryKey: 'posts',
+  queryFn: fetchPosts,
+  staleTime: Infinity,
+})
+```
+
+### queryCache.invalidateQueries()
+
+The `client.invalidateQueries()` method will now only invalidate queries.
```

Comment:
I thought `invalidateQueries` was still going to be "smart", eg. by refetching active queries and marking inactive ones as stale. Is there no way to do that now?

Reply:
The idea behind this was to allow users a very efficient way to do:
```js
client.invalidateQueries('posts', { active: false, inactive: true })
client.refetchQueries('posts', { active: true, inactive: false })
```

Reply:
What I like about the separation is that it is really clear what each method does. The `invalidateQueries()` method marks queries as invalid and the `refetchQueries()` method refetches queries. Most people probably just want to refetch?

Reply:
I thought so too. But then you have a long season of cache behind you and trigger refetches for a query key prefix for example, you run the risk of triggering a massive amount of requests from all of the inactive cache entries. However you also want to mark those as stale.

Reply:
But I see your point. Essentially you would suggest using both with the correct inactive and active filters, right?

Reply:
In this implementation `refetchQueries()` only fetches active queries by default to prevent over-fetching. But it does not also invalidate inactive queries when specifying a prefix. Think the `refetchQueries()` concept is simpler and more easy to grasp for users, but they might "forget" that when specifying a prefix, there could be inactive queries that might need to be marked as invalid. Hmmm

Reply:
I love it when you go "Hmmmm..." Good things usually come soon after lol.